File: core/src/main/java/xtdb/vector/BoxWriter.java
package xtdb.vector;

import clojure.lang.Keyword;
import org.jetbrains.annotations.Nullable;

import java.nio.ByteBuffer;

abstract class BoxWriter implements IValueWriter {
    abstract IValueWriter box();

    @Override
    public void writeNull() {
        box().writeNull();
    }

    @Override
    public void writeBoolean(boolean v) {
        box().writeBoolean(v);
    }

    @Override
    public void writeByte(byte v) {
        box().writeByte(v);
    }

    @Override
    public void writeShort(short v) {
        box().writeShort(v);
    }

    @Override
    public void writeInt(int v) {
        box().writeInt(v);
    }

    @Override
    public void writeLong(long v) {
        box().writeLong(v);
    }

    @Override
    public void writeFloat(float v) {
        box().writeFloat(v);
    }

    @Override
    public void writeDouble(double v) {
        box().writeDouble(v);
    }

    @Override
    public void writeBytes(ByteBuffer v) {
        box().writeBytes(v);
    }

    @Override
    public void writeObject(@Nullable Object obj) {
        box().writeObject(obj);
    }

    @Override
    public IValueWriter legWriter(Keyword leg) {
        return box().legWriter(leg);
    }
}

File: core/src/main/kotlin/xtdb/arrow/BoxWriter.kt
package xtdb.arrow

import java.nio.ByteBuffer

internal abstract class BoxWriter : ValueWriter {
    abstract fun box(): ValueBox

    override fun writeNull() {
        box().writeNull()
    }

    override fun writeBoolean(v: Boolean) {
        box().writeBoolean(v)
    }

    override fun writeByte(v: Byte) {
        box().writeByte(v)
    }

    override fun writeShort(v: Short) {
        box().writeShort(v)
    }

    override fun writeInt(v: Int) {
        box().writeInt(v)
    }

    override fun writeLong(v: Long) {
        box().writeLong(v)
    }

    override fun writeFloat(v: Float) {
        box().writeFloat(v)
    }

    override fun writeDouble(v: Double) {
        box().writeDouble(v)
    }

    override fun writeBytes(v: ByteBuffer) {
        box().writeBytes(v)
    }

    override fun writeObject(obj: Any?) {
        box().writeObject(obj)
    }

    fun legWriter(leg: String): ValueWriter {
        return box().legWriter(leg)
    }
}
